From bd5f58ca8950962a52b0fb9d566b5aac146242b2 Mon Sep 17 00:00:00 2001
From: Ramanathan Choodamani <quic_rchoodam@quicinc.com>
Date: Tue, 14 May 2024 22:48:18 -0700
Subject: [PATCH] wifi: mac80211: Avoid valid links check for non MLO clients 
 in rx slow path

When a non MLO client sends UL traffic, the packets in slow path are
dropped as the valid_links bitmap will be 0 and not delivered to the
stack.
Add a sanity check to ensure these packets are not dropped.

Patch-work: None
Patch dependency: 784-mac80211-fix-skb-and-sdata-mismatch.patch
Signed-off-by: Ramanathan Choodamani <quic_rchoodam@quicinc.com>
---
 net/mac80211/rx.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/net/mac80211/rx.c b/net/mac80211/rx.c
index 83e8828..49cff81 100644
--- a/net/mac80211/rx.c
+++ b/net/mac80211/rx.c
@@ -4416,6 +4416,9 @@ static void ieee80211_invoke_rx_handlers(struct ieee80211_rx_data *rx)
 static bool
 ieee80211_rx_is_valid_sta_link_id(struct ieee80211_sta *sta, u8 link_id)
 {
+	if (!sta->mlo && !link_id)
+		return true;
+
 	return !!(sta->valid_links & BIT(link_id));
 }
 
-- 
2.25.1

